### PR TITLE
Add Morphology::diff method that is like operator=! but verbose

### PR DESCRIPTION
**Pull request by wizmer**

### CODE REVIEW

File: src/properties.cpp [hunk 1]
```diff
@@ -209,73 +178,108 @@ bool compare(const T& el1, const T& el2, const std::string& name, bool verbose_)
...
 }
 
 bool CellLevel::operator!=(const CellLevel& other) const
 {
-    return !(this->operator==(other));
+    return diff(other, false);
+}
+
+
+MitochondriaPointLevel::MitochondriaPointLevel(
```
Comment (wizmer) [inline src/properties.cpp]: This is moved code, not new code

File: src/mut/morphology.cpp [hunk 2]
```diff
@@ -411,38 +411,32 @@ void Morphology::write(const std::string& filename)
         LBTHROW(UnknownFileType(_err.ERROR_WRONG_EXTENSION(filename)));
 }
 
-bool Morphology::_compare(const Morphology& other, bool verbose) const
```
Comment (arsenius7) [inline src/mut/morphology.cpp]: `compare` looked just as good to me (I would also make it a free method, but that's probably a matter of taste).

File: src/mut/section.cpp [hunk 3]
```diff
@@ -203,57 +202,69 @@ std::shared_ptr<Section> Section::appendSection(
...
 }
 
-bool Section::_compare(const Section& other, bool verbose) const
+bool Section::diff(const Section& other, bool verbose) const
 {
     if (this->type() != other.type()) {
         if (verbose)
-            std::cout << "Section type differ" << std::endl;
-        return false;
+            std::cout << "Reason: section type differ" << std::endl;
```
Comment (arsenius7) [inline src/mut/section.cpp]: maybe `std::cerr`?

File: src/mut/section.cpp [hunk 4]
```diff
@@ -91,12 +91,11 @@ upstream_iterator Section::upstream_end() const
...
 std::ostream& operator<<(std::ostream& os, Section& section)
 {
-    os << "id: " << section.id() << std::endl;
-    os << dumpPoints(section.points());
+    ::operator<<(os, section);
     return os;
 }
 
-std::ostream& operator<<(std::ostream& os, std::shared_ptr<Section> sectionPtr)
+std::ostream& operator<<(std::ostream& os, const std::shared_ptr<Section> sectionPtr)
```
Comment (arsenius7) [inline src/mut/section.cpp]: it's a copy anyway at the moment, so `const` does not change much

File: src/mut/section.cpp (hunk 3 above)
Comment (wizmer) [reply src/mut/section.cpp]: I am not sure. The message is what is expected to be print out, it is not really an error.

File: src/mut/morphology.cpp (hunk 2 above)
Comment (wizmer) [reply src/mut/morphology.cpp]: good point, I'll change that

File: src/mut/section.cpp (hunk 3 above)
Comment (arsenius7) [reply src/mut/section.cpp]: `stderr` is not necessarily for errors, but for additional "diagnostics" as well.
But now that I took a look at some `diff`-ing tools, all of them use `stdout`, so nvm.

**Comment by wizmer** [comment]
hawk eyes

File: setup.py [hunk 5]
```diff
@@ -44,6 +44,7 @@ def build_extension(self, ext, cmake):
         cmake_args = ['-DCMAKE_LIBRARY_OUTPUT_DIRECTORY=' + extdir,
                       '-DPYTHON_EXECUTABLE=' + sys.executable]
 
+        self.debug = True
```
Comment (arsenius7) [inline setup.py]: leftover?

File: include/morphio/properties.h [hunk 6]
```diff
@@ -106,40 +125,30 @@ struct MitochondriaPointLevel
...
+                           SectionRange range);
 
     MitochondriaPointLevel(
         std::vector<uint32_t> sectionId,
         // relative pathlength between the current points
         // and the start of the neuronal section
         std::vector<MitoPathLength::Type> relativePathLengths,
         std::vector<MitoDiameter::Type> diameters);
+
+    bool diff(const MitochondriaPointLevel& other, LogLevel verbose) const;
```
Comment (arsenius7) [inline include/morphio/properties.h]: Maybe it needs to be renamed from `verbose` to `logLevel` now.
Also, as discussed verbally, another option is an environment variable controlling log verbosity for all `MorphIO` methods.

File: src/mut/section.cpp [hunk 7]
```diff
@@ -203,57 +203,23 @@ std::shared_ptr<Section> Section::appendSection(
...
-    return !_compare(other, false);
+    return diff(*this, other, LogLevel::ERROR);
 }
 
 } // end namespace mut
 } // end namespace morphio
+
+std::ostream& operator<<(std::ostream& os, const morphio::mut::Section& section)
+{
+    auto points = section.points();
```
Comment (arsenius7) [inline src/mut/section.cpp]: I guess this can not be empty, right?

File: src/properties.cpp [hunk 8]
```diff
@@ -80,19 +49,19 @@ PointLevel& PointLevel::operator=(const PointLevel& other)
...
 bool compare(const std::vector<T>& vec1, const std::vector<T>& vec2,
-    const std::string& name, bool verbose_)
+    const std::string& name, LogLevel verbose)
 {
     if (vec1 == vec2)
         return true;
 
     if (vec1.size() != vec2.size()) {
-        if (verbose_)
+        if (verbose > LogLevel::ERROR)
```
Comment (arsenius7) [inline src/properties.cpp]: with global log level we can move it to LBERROR macro

**Pull request merged**